internal/btf/syscalls.go

```diff
@@ -0,0 +1,60 @@
+package btf
+
+import (
+	"errors"
+	"fmt"
+	"unsafe"
+
+	"github.com/cilium/ebpf/internal"
+	"github.com/cilium/ebpf/internal/unix"
+)
+
+type bpfBTFInfo struct {
+	btf       internal.Pointer
+	btfSize   uint32
+	id        uint32
+	name      internal.Pointer
+	nameLen   uint32
+	kernelBTF uint32
+}
+
+func bpfGetBTFInfoByFD(fd *internal.FD, btf, name []byte) (*bpfBTFInfo, error) {
+	info := bpfBTFInfo{
+		btf:     internal.NewSlicePointer(btf),
+		btfSize: uint32(len(btf)),
+		name:    internal.NewSlicePointer(name),
+		nameLen: uint32(len(name)),
+	}
+	if err := internal.BPFObjGetInfoByFD(fd, unsafe.Pointer(&info), unsafe.Sizeof(info)); err != nil {
+		return nil, fmt.Errorf("can't get program info: %w", err)
+	}
+
+	return &info, nil
+}
+
+// Everything below here is redefined to prevent an import cycle
+type bpfGetFDByIDAttr struct {
```

Comment:
I think that would become a breaking change or being a bit clunky.
The implementation in `ebpf` can return `ebpf.ErrNotExist`. `internal` cannot import `ebpf` without creating an import cycle, so it cannot return `ebpf.ErrNotExist`. We can't just move `ErrNotExist` into `internal` without a breaking change because it was previously exported from `ebpf` and available to users.
We could define a new error inside of `internal` without removing the one in `ebpf`, but then any functions that currently wrap `bpfObjGetFDByID()` would need to translate `internal.ErrNotExist` into `ebpf.ErrNotExist` before returning an error to the user, or, again, breaking change.
Thoughts?

Reply:
Have you pulled master recently? This is what the function looks like now:
```
func bpfObjGetFDByID(cmd internal.BPFCmd, id uint32) (*internal.FD, error) {
	attr := bpfGetFDByIDAttr{
		id: id,
	}
	ptr, err := internal.BPF(cmd, unsafe.Pointer(&attr), unsafe.Sizeof(attr))
	return internal.NewFD(uint32(ptr)), err
}
```

`ebpf.ErrNotExist` is also deprecated.